pkg/utils/authcheck/authtype.go

```diff
@@ -0,0 +1,139 @@
+	authTypeForWorkloadIdentity, workloadIdentityErr := getAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args)
+	authTypeForSecret, secretErr := getAuthTypeForSecret(ctx, secretLister, args)
+	if authTypeForWorkloadIdentity != "" {
+		return authTypeForWorkloadIdentity, nil
+	} else if authTypeForSecret != "" {
+		return authTypeForSecret, nil
+	} else {
+		workloadIdentityError := fmt.Errorf("when checking Kubernetes Service Account %s, got error: %w", args.ServiceAccountName, workloadIdentityErr)
+		secretError := fmt.Errorf("when checking Kubernetes Secret %s, got error: %w", args.Secret.Name, secretErr)
+		return "", fmt.Errorf("authentication is not configured, %s, %s", workloadIdentityError.Error(), secretError.Error())
+	}
+}
+
+// GetAuthTypeForSources will get authType for Sources.
+func GetAuthTypeForSources(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister, args AuthTypeArgs) (AuthTypes, error) {
+	// For AuthTypeArgs from Sources, either ServiceAccountName or Secret will be empty,
+	// because of the IdentitySpec validation from Webhook.
+
+	// For AuthTypeArgs from Sources which has serviceAccountName.
+	if args.ServiceAccountName != "" {
+		authType, err := getAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args)
+		if err != nil {
+			return authType, fmt.Errorf("using Workload Identity for authentication configuration: %w", err)
+		}
+		return authType, nil
+	}
+
+	// For AuthTypeArgs from Sources which has secret.
+	if args.Secret != nil {
+		// Sources' secrets are not further checked.
+		// In most cases, sources don't live in the control plane's namespace,
+		// and the controller doesn't have the permission to check their secrets.
+		return Secret, nil
+	}
+
+	return "", errors.New("invalid AuthTypeArgs, neither ServiceAccountName nor Secret are provided")
+}
+
+func getAuthTypeForWorkloadIdentity(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister,
+	args AuthTypeArgs) (AuthTypes, error) {
+	kServiceAccount, err := serviceAccountLister.ServiceAccounts(args.Namespace).Get(args.ServiceAccountName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			return "", fmt.Errorf("can't find Kubernetes Service Account %s",
+				args.ServiceAccountName)
+		}
+		return "", fmt.Errorf("error getting Kubernetes Service Account: %w", err)
+	} else if kServiceAccount.Annotations[resources.WorkloadIdentityKey] != "" {
+		return WorkloadIdentityGSA, nil
+	}
+	// Once workload-identity new gen lands, we should also include the annotation check for it.
+	return "", fmt.Errorf("the Kubernetes Service Account %s does not have the required annotation", args.ServiceAccountName)
+}
+
+func getAuthTypeForSecret(ctx context.Context, secretLister corev1listers.SecretLister, args AuthTypeArgs) (AuthTypes, error) {
+	// Controller doesn't have the permission to check the existence of a secret in namespaces
+	// other than the control plane's namespace.
```

Comment:
Awesome! Thanks for the comment, as I won't remember what you said the next time I'm reading the code.